<a href="https://colab.research.google.com/github/KOMBOU12/Marius/blob/main/Matrix_Completion.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Le but est d'implémenter le code présent dans ce lien avec un jeu de donnée de mon choix : https://colab.research.google.com/drive/1-KStBJ_gHRjRrAwaMj6cBZLymqiEjpnm?usp=sharing

In [17]:
pip install matrix-completion

In [18]:
import numpy as np
from matrix_completion import svt_solve, calc_unobserved_rmse

U = np.random.randn(20, 5)
V = np.random.randn(15, 5)
R = np.random.randn(20, 15) + np.dot(U, V.T)

mask = np.round(np.random.rand(20, 15))
R_hat = svt_solve(R, mask)

print("RMSE:", calc_unobserved_rmse(U, V, R_hat, mask))

RMSE: 2.1877459684965315


Le code génère une matrice de rang faible R avec des valeurs manquantes, utilise l'algorithme **SVT pour imputer les valeurs manquantes**, et évalue la qualité de l'imputation en calculant la RMSE sur les valeurs non observées.

**On refait le code avec de nouvelles données**

On commence par générer des matrices aléatoires $U_{new}$ et $V_{new}$. Puis on crée une matrice de rotation R en ajoutant du bruit gaussien à $U_{new}.V_{new}^{T}$

In [19]:
import numpy as np
from matrix_completion import svt_solve, calc_unobserved_rmse

# Generation de nouvelles matrices U, V, et R
U_new = np.random.randn(25, 7)  # U est maintenant 25x7
V_new = np.random.randn(18, 7)  # V est maintenant 18x7
R_new = np.random.randn(25, 18) + np.dot(U_new, V_new.T)  # R est maintenant 25x18 avec une faible structure de rang
R_new

array([[ 4.02900220e+00, -3.16163809e+00, -4.87931625e+00,
        -7.88485137e+00,  3.13425929e+00, -3.49392450e-01,
         1.25236547e+00, -1.60823649e+00, -1.75660303e+00,
        -1.19160523e+00,  8.48964151e-01,  3.27467245e-01,
         4.05301311e+00, -2.75124425e+00,  1.39794066e+00,
        -6.31872311e-01,  7.27535345e-01,  1.39518917e+00],
       [ 4.89266722e+00, -2.18406268e-01,  1.99281178e+00,
        -6.45341911e+00, -3.72950557e+00,  1.99492879e+00,
         3.07791433e+00, -5.12286232e+00,  3.67817184e+00,
        -2.15015807e+00,  2.82154216e+00,  2.50056540e+00,
        -3.72382712e+00,  4.55368143e-01,  2.88077663e-01,
        -1.20399323e+00, -3.96261736e-03,  2.57338654e+00],
       [-4.28220933e+00,  9.52012872e-01, -3.18321006e-01,
         6.41913939e+00,  8.17016798e-01, -8.20706808e-01,
        -5.64079467e-01,  2.28740999e+00, -2.30403484e+00,
         1.79012916e+00,  3.21477656e-02,  1.54034036e+00,
         2.33466666e-01, -3.18889334e+00, -2.50442996e

Un masque binaire est appliqué sur R pour simuler des valeurs observées et non observées. La fonction np.round(np.random.rand(20, 15)) crée une matrice de la même taille que $R$ avec des valeurs aléatoires de 0 ou 1, où 1 indique une donnée observée et 0 une donnée manquante.


In [20]:
# Creation d'un nouveau masque pour les valeurs manquantes
mask_new = np.round(np.random.rand(25, 18))

In [21]:
# Application de la SVT pour compléter la matrice
R_hat_new = svt_solve(R_new, mask_new)

# Calcul du RMSE pour les valeurs pas observées
rmse_new = calc_unobserved_rmse(U_new, V_new, R_hat_new, mask_new)

# le nouveau résultat du RMSE
print("New RMSE:", rmse_new)

New RMSE: 1.8703451944299176



$R_{new}$ est de taille 25x18 avec des valeurs fluctuant autour de -10 à 10. La valeur RMSE évalue la différence moyenne au carré entre les valeurs estimées (ou prédites) et les valeurs réelles. Dans ce contexte, une RMSE de 1.87 peut être considérée comme **raisonnablement bonne** , car cela signifierait que l'erreur représente une petite fraction de la variabilité totale des données.

Pour **améliorer la précision du modèle**, on peut envisager, d'utiliser plus de données d'entraînement, ou d'améliorer la qualité des données utilisées.